ocaml-lsp-server/src/workspace_symbol.ml

```diff
@@ -0,0 +1,216 @@
+  let sourcefile =
+    root_uri ^ "/" ^ Option.value ~default:"" cmt.cmt_sourcefile
+  in
+  let signatures = cmt_sign cmt in
+  signatures |> List.map ~f:(symbol_of_signature_item ~sourcefile)
+
+type cm_file =
+  | Cmt of string
+  | Cmti of string
+
+let string_of_cm cm =
+  match cm with
+  | Cmt f
+  | Cmti f ->
+    f
+
+module Cm_files = Map.Make (struct
+  type t = string
+
+  let compare a b = Stdlib.compare a b |> Ordering.of_int
+
+  let to_dyn t = Dyn.String t
+end)
+
+let find_dirfiles dirs =
+  let choose_file f1 f2 =
+    match (f1, f2) with
+    | (Cmt _ as f), _
+    | _, (Cmt _ as f) ->
+      f
+    | (Cmti _ as f), Cmti _ -> f
+  in
+  let split_filename file =
+    let maybe_index = String.rindex_opt file '.' in
+    let len = String.length file in
+    match maybe_index with
+    | None -> (file, "")
+    | Some index ->
+      let modul = String.capitalize_ascii (String.sub file ~pos:0 ~len:index) in
+      let ext =
+        String.lowercase_ascii
+          (String.sub file ~pos:(index + 1) ~len:(len - index - 1))
+      in
+      (modul, ext)
+  in
+  List.fold_left
+    ~f:(fun files dir ->
+      Sys.readdir dir
+      |> Array.fold_left
+           ~f:(fun acc file ->
+             let modul, ext = split_filename file in
+             let path = dir ^ "/" ^ file in
+             match ext with
+             | "cmt" -> Cm_files.set acc modul (Cmt path)
+             | "cmti" -> (
+               let current_file = Cm_files.find acc modul in
+               let cmti_file = Cmti path in
+               match current_file with
+               | None -> Cm_files.set acc modul cmti_file
+               | Some current_file ->
+                 Cm_files.set acc modul (choose_file current_file cmti_file))
+             | _ -> acc)
+           ~init:files)
+    ~init:Cm_files.empty dirs
+  |> Cm_files.values |> List.map ~f:string_of_cm
+
+type error = No_build_dir of string
+
+let run ({ query; _ } : WorkspaceSymbolParams.t) (rootUri : Uri.t option) =
+  let rootPath =
+    Option.map rootUri ~f:Uri.to_path |> Option.value ~default:(Sys.getcwd ())
+  in
+  let build_dir = find_build_dir rootPath in
+  match build_dir with
+  | None -> Error (No_build_dir rootPath)
+  | Some build_dir -> (
+    let subdirs = subdirs build_dir in
+    let all_files = find_dirfiles subdirs in
+    let load_file file =
+      let maybe_index = String.rindex_opt file '.' in
+      match maybe_index with
+      | None -> []
+      | Some i -> (
+        let len = String.length file in
+        let ext =
+          String.lowercase_ascii
+            (String.sub file ~pos:(i + 1) ~len:(len - i - 1))
+        in
+        match ext with
+        | "cmt"
+        | "cmti" ->
+          read_cmt rootPath file
+        | _ -> [])
+    in
+    let all_symbols = List.map ~f:load_file all_files |> List.flatten in
+    match query with
+    | "" -> Ok all_symbols
+    | query ->
+      let reg_exp = Str.regexp_string query in
```

Comment:
Unfortunately, `Str` will not work for us (it's too hard to use correctly in a multi threaded context). Use `Re` instead.